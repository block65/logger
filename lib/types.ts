import { AsyncLocalStorage } from 'async_hooks';
import type pino from 'pino';
import type { MixinFn, MixinFnWithData } from './mixins.js';

export type Falsy = false | undefined | null;

export enum LogLevelNumbers {
  Fatal = 60,
  Error = 50,
  Warn = 40,
  Info = 30,
  Debug = 20,
  Trace = 10,
  Silent = 0,
}

export interface LogDescriptor {
  level: LogLevelNumbers;
  time: number;

  msg?: string;
  name?: string | undefined;

  // maye not be available depending on platform/cli
  pid?: number;
  hostname?: string;

  [key: string]: unknown;
}

type RemoveIndex<T> = {
  [K in keyof T as {} extends Record<K, 1> ? never : K]: T[K];
};

// pino tacks on a nasty `Record<string, any>` to its logger type
// we try to remove it here
export type BaseLogger = RemoveIndex<pino.Logger>;

export interface Logger extends BaseLogger {
  als: AsyncLocalStorage<AlsContext>;
}

export interface AlsContext {
  id: string;
  context?: Record<string, unknown>;
}

export interface AlsContextOutput {
  _contextId: AlsContext['id'];
  _context?: AlsContext['context'];
}

export type ComputePlatform = 'gcp-cloudrun' | 'aws-lambda' | 'aws';

export interface CreateLoggerOptions
  extends Omit<pino.LoggerOptions, 'mixin' | 'prettyPrint' | 'prettifier'> {
  pretty?: boolean;
  traceCaller?: boolean;
  platform?: ComputePlatform;
  mixins?: (MixinFnWithData | MixinFn | Falsy)[];
  transports?: pino.TransportMultiOptions;
}

export type CreateLoggerOptionsWithoutTransports = Omit<
  CreateLoggerOptions,
  'pretty' | 'transports'
>;

export interface CreateCliLoggerOptions
  extends Omit<CreateLoggerOptionsWithoutTransports, 'platform'> {
  color?: boolean;
}
