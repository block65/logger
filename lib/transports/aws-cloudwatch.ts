import { once } from 'events';
import build from 'pino-abstract-transport';
import SonicBoom, { SonicBoomOpts } from 'sonic-boom';
import { LogDescriptor, LogLevelNumbers } from '../types.js';
import { stringifyUndefined } from '../utils.js';

// must be serialiazable so it can get passed to a worker
export interface CloudwatchTransportOptions extends SonicBoomOpts {}

// https://github.com/aws/aws-lambda-nodejs-runtime-interface-client/blob/c31c41ffe5f2f03ae9e8589b96f3b005e2bb8a4a/src/utils/LogPatch.ts#L10
type AWsLevelNames = 'FATAL' | 'ERROR' | 'WARN' | 'INFO' | 'DEBUG' | 'TRACE';

const levelNumberToCloudwatchStringMap = new Map<
  LogLevelNumbers,
  AWsLevelNames
>([
  [LogLevelNumbers.Fatal, 'FATAL'],
  [LogLevelNumbers.Error, 'ERROR'],
  [LogLevelNumbers.Warn, 'WARN'],
  [LogLevelNumbers.Info, 'INFO'],
  [LogLevelNumbers.Debug, 'DEBUG'],
  [LogLevelNumbers.Trace, 'TRACE'],
  [LogLevelNumbers.Silent, 'TRACE'],
]);

function createCloudwatchTransformer() {
  return (log: LogDescriptor): string => {
    const {
      msg,
      // eslint-disable-next-line @typescript-eslint/naming-convention
      _contextId,
      // eslint-disable-next-line @typescript-eslint/naming-convention
      _context,
      time,
      ...rest
    } = log;

    const restStr =
      Object.keys(rest).length > 0
        ? JSON.stringify({ ...stringifyUndefined(rest), ctx: _context })
        : '';

    return [
      new Date(time).toISOString(),
      _contextId || '',
      levelNumberToCloudwatchStringMap.get(log.level),
      msg || '',
      restStr,
    ].join('\t');
  };
}

export function cloudwatchTransport(options: CloudwatchTransportOptions = {}) {
  const transformer = createCloudwatchTransformer();

  // SonicBoom is necessary to avoid loops with the main thread.
  // It is the same of pino.destination().
  const destination = new SonicBoom(options);

  // await once(destination, 'ready');

  return build(
    async (source) => {
      // eslint-disable-next-line no-restricted-syntax
      for await (const obj of source) {
        const toDrain = !destination.write(`${transformer(obj)}\n`);
        // This block will handle backpressure
        if (toDrain) {
          await once(destination, 'drain');
        }
      }
    },
    {
      async close(err?: Error) {
        if (err) {
          console.warn(err);
        }
        destination.end();

        await once(destination, 'close');
      },
    },
  );
}

export default cloudwatchTransport;
