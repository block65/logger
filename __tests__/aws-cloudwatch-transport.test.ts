import {
  afterEach,
  beforeEach,
  describe,
  expect,
  jest,
  test,
} from '@jest/globals';
import { cloudwatchTransport } from '../lib/transports/aws-cloudwatch.js';
import { LogLevelNumbers } from '../lib/types.js';
import { generateTmpFilenameAndReader, writeLogsToStream } from './helpers.js';

describe('AWS Cloudwatch', () => {
  beforeEach(() => {
    jest.useFakeTimers('modern');
    jest.setSystemTime(new Date('2009-02-13T23:31:30.000Z'));
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  test('Transport', async () => {
    const [dest, getLogs] = await generateTmpFilenameAndReader();
    const transport = await cloudwatchTransport({ dest });

    writeLogsToStream(
      transport,
      {
        level: LogLevelNumbers.Trace,
        time: Date.now(),
        msg: 'trace',
        random: {
          bool: true,
          str: 'hello',
          number: 12,
          arr: [1, 2, 3],
        },
      },
      {
        level: LogLevelNumbers.Debug,
        time: Date.now(),
        msg: 'Debug',
      },
      {
        level: LogLevelNumbers.Info,
        time: Date.now(),
        msg: 'Info',
      },
      {
        level: LogLevelNumbers.Warn,
        time: Date.now(),
        msg: 'Warn',
      },
      {
        level: LogLevelNumbers.Error,
        time: Date.now(),
        msg: 'Error',
        err: {
          stack: ['stuff', 'code'],
        },
      },
      {
        level: LogLevelNumbers.Fatal,
        time: Date.now(),
        msg: 'Fatal',
        err: {
          stack: ['stuff', 'code'],
        },
      },
    );

    await expect(getLogs()).resolves.toMatchSnapshot();
  });
});
