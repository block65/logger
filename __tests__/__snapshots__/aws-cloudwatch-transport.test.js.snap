// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AWS Cloudwatch Transport 1`] = `
"2009-02-13T23:31:30.000Z		TRACE	trace	{\\"pid\\":1,\\"hostname\\":\\"test\\",\\"random\\":{\\"bool\\":true,\\"str\\":\\"hello\\",\\"number\\":12,\\"arr\\":[1,2,3]}}
2009-02-13T23:31:30.000Z		DEBUG	Debug	
2009-02-13T23:31:30.000Z		INFO	Info	
2009-02-13T23:31:30.000Z		WARN	Warn	
2009-02-13T23:31:30.000Z		ERROR	Error	{\\"stack\\":[\\"stuff\\",\\"code\\"]}
2009-02-13T23:31:30.000Z		FATAL	Fatal	{\\"stack\\":[\\"stuff\\",\\"code\\"]}"
`;
